### PR TITLE
guest_memory: Fix region access violation

### PR DESCRIPTION
**Pull request by bjzhjing**
While call try_access from read_from, in try_access, it does not
check if start + len will cross over the end of the current region.
If it will, once pass the region to the closure, where the region
be converted to slice, and data is picked up there, from the start
to the end = start + len, then, region access violation will happen,
since start + len is out of range of the slice.

Get the len in a differen way in try_access to ensure it will not
cross over the region to avoid such error.

Signed-off-by: Cathy Zhang <cathy.zhang@intel.com>

### CODE REVIEW

File: src/guest_memory.rs [hunk 1]
```diff
@@ -295,8 +295,8 @@ pub trait GuestMemory {
         let mut total = 0;
         while let Some(region) = self.find_region(cur) {
             let start = region.to_region_addr(cur).unwrap();
-            let cap = region.len() as usize;
-            let len = std::cmp::min(cap, count - total);
+            let size = region.len() as usize - start.raw_value() as usize;
```
Comment (jiangliu) [inline src/guest_memory.rs]: Sorry, at a second glance found it's unsafe to cast region.len() and start.raw_value() as usize.
usize is 32bit with a 32bit VMM but region.len() and start.raw_value() are both u64.

**Comment by bjzhjing** [comment]
@chao-p  Yes, I encounter this error while adding hugepages support, in the case with hugepagesize=2M, it fails to write elf image to memory. After add this code change, elf image could be loaded.

**Comment by bonzini** [comment]
Looks good but it needs a test case. Also the name of the variable "cap" is better: `size` is accurate in the buggy code but not in the fixed one.

File: src/guest_memory.rs (hunk 1 above)
Comment (bonzini) [reply src/guest_memory.rs]: Yes, this must be `GuestUsize`.

Comment (bjzhjing) [reply src/guest_memory.rs]: It seems this is not the only case in guest_memory.rs regarding the potential truncate issue, I suggest to have a separate PR opened to handle such issue.

**Comment by bjzhjing** [comment]
@bonzini I thought "cap" represents the capacity of the region, but here the parameter is only the length/size of the sub-set data in the region. Anyway, it's changed back to be "cap", and a test case is added for this change :-P

File: src/guest_memory.rs (hunk 1 above)
Comment (bonzini) [reply src/guest_memory.rs]: Still, this should use `GuestUsize`.

File: src/guest_memory.rs [hunk 2]
```diff
@@ -295,8 +295,8 @@ pub trait GuestMemory {
         let mut total = 0;
         while let Some(region) = self.find_region(cur) {
             let start = region.to_region_addr(cur).unwrap();
-            let cap = region.len() as usize;
-            let len = std::cmp::min(cap, count - total);
+            let cap = region.len() - start.raw_value();
+            let len = std::cmp::min(cap as usize, count - total);
```
Comment (jiangliu) [inline src/guest_memory.rs]: There is still a risk that cap is bigger than usize on 32-bit platforms.
How about:
            let cap = region.len() - start.raw_value();
            let len = std::cmp::min(cap, (count - total) as GuestUsize);	   
           match f(total, len as usize, start, region) {

Comment (bonzini) [reply src/guest_memory.rs]: Yes, it's better. @bjzhjing can you adjust?

Comment (bjzhjing) [reply src/guest_memory.rs]: Sure, change is done!

**Comment by jiangliu** [comment]
@bjzhjing hi, this PR has some conflicts with the latest master branch, so could you help to rebase the PR?

**Comment by bjzhjing** [comment]
@jiangliu Rebase to the newest version, and submit the commit again.

**Comment by bjzhjing** [comment]
Is it possible to merge the change and close the PR ASAP, since Gerry and Paolo have approve the change previously. There is no update during the latest rebase.

**Comment by jiangliu** [comment]
> Is it possible to merge the change and close the PR ASAP, since Gerry and Paolo have approve the change previously. There is no update during the latest rebase.

Still need another approver:(

**Pull request merged**